### PR TITLE
[SQL] Add an exception for analysis errors.

### PR DESCRIPTION
**Pull request by marmbrus**
Also start from the bottom so we show the first error instead of the top error.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 1]
```diff
@@ -33,6 +33,11 @@ import org.apache.spark.sql.types.IntegerType
 object SimpleAnalyzer extends Analyzer(EmptyCatalog, EmptyFunctionRegistry, true)
 
 /**
+ * Thrown when a query fails to analyze, usually because the query itself is invalid.
+ */
+class AnalysisException(message: String) extends Exception(message) with Serializable
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: can we move this to org.apache.sql so we can document it?

Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Good point.

**Comment by scwf** [comment]
retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 2]
```diff
@@ -81,16 +82,18 @@ class Analyzer(catalog: Catalog,
    */
   object CheckResolution extends Rule[LogicalPlan] {
     def apply(plan: LogicalPlan): LogicalPlan = {
-      plan.transform {
+      plan.transformUp {
         case p if p.expressions.exists(!_.resolved) =>
-          throw new TreeNodeException(p,
-            s"Unresolved attributes: ${p.expressions.filterNot(_.resolved).mkString(",")}")
+          val missing = p.expressions.filterNot(_.resolved).map(_.prettyString).mkString(",")
```
Comment (scwf) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: prettyString is added in #4436, which have not in...

**Comment by cloud-fan** [comment]
Should we throw `AnalysisException` instead of `sys.error` at [here](https://github.com/marmbrus/spark/blob/analysisException/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala#L322)?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 2 above)
Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Yeah, thanks.  I had a giant branch that I broke up into 5 different PRs.  Will retest once that is merged.

**Comment by cloud-fan** [comment]
Sorry I forgot that... Once you throw `AnalysisException` at  [here](https://github.com/marmbrus/spark/blob/analysisException/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala#L322), you also need to update the test case [SPARK-5278](https://github.com/apache/spark/blob/master/sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveResolutionSuite.scala#L43) to expect `AnalysisException`.

**Pull request closed without merging**